### PR TITLE
[composite compliance] forward_ad: linear

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #69991

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -47,7 +47,8 @@ Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Ten
   auto output = at::matmul(input, weight.t());
   if (bias->defined()) {
     // for composite compliance use out-of-place version of `add`
-    if (isTensorSubclassLike(*bias)) {
+    if (isTensorSubclassLike(*bias) ||
+        (*bias)._fw_grad(/*level*/ 0).defined()) {
```
Comment (kshitij12345) [inline aten/src/ATen/native/Linear.cpp:51]: Is this the correct way to check if forward mode AD is being used?

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -47,7 +47,8 @@ Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Ten
   auto output = at::matmul(input, weight.t());
   if (bias->defined()) {
     // for composite compliance use out-of-place version of `add`
-    if (isTensorSubclassLike(*bias)) {
+    if (isTensorSubclassLike(*bias) ||
```
Comment (zou3519) [inline aten/src/ATen/native/Linear.cpp:50]: @kshitij12345 and I discussed a bit offline, but just to record for posterity: The probably with F.linear and composite compliance and forward-ad happens in the case of `F.linear(input=Tensor, weight=Tensor, bias=DualTensor[primal: Tensor, Tangent: CCTensor])`. I'm not totally sure why we need to "check if forward mode AD is being used", but I think there's an open question on if the "if Tensor is a dual tensor" should also become a part of the isTensorSubclassLike check.

Comment (kshitij12345) [reply aten/src/ATen/native/Linear.cpp:50]: For `F.linear(input=Tensor, weight=Tensor, bias=DualTensor[primal: Tensor, Tangent: CCTensor])`, if we take the inplace branch
https://github.com/pytorch/pytorch/blob/5d5783caea193a2762e6181b1c5eb6718e4b8f88/aten/src/ATen/native/Linear.cpp#L53-L55

Then `output` and its tangent are just Tensors while bias is a Tensor and bias's tangent is `CCT`

Looking at the forward_ad path for `add_` generated in `VariableType3.cpp`,
```
at::Tensor & add__Tensor(c10::DispatchKeySet ks, at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha)
```

The part where gradients are computed,
```
if (_any_has_forward_grad_self) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::zeros(self_tensor.sizes(), self_tensor.options());
      auto other_t_raw = toNonOptFwGrad(other);
      auto other_tensor = toNonOptTensor(other);
      auto other_t = (other_t_raw.defined() || !other_tensor.defined())
        ? other_t_raw : at::_efficientzerotensor(other_tensor.sizes(), other_tensor.options());
      self_t = GradMode::is_enabled() ? self_t.clone() : self_t;
      self_new_fw_grad_opt = self_t_raw.defined() ? self_t_raw.copy_(self_t + maybe_multiply(other_t, alpha)) : self_t + maybe_multiply(other_t, alpha);
  }
```

At last line, we know see that `self_t_raw` is a normal Tensor while the value being copied is `CCTensor` and hence the test fails.

Comment (kshitij12345) [reply aten/src/ATen/native/Linear.cpp:50]: In `test_forward_ad`, we always generate tangents for all valid `args` and `kwargs` (Tensors with requires_grad set)
https://github.com/pytorch/pytorch/blob/8d4c8df33a58dc5f905dfdcee1cd124a79aaf2d8/torch/testing/_internal/composite_compliance.py#L469-L472

Looking at the gradient computation snippet above, we never take the path for the case when `self_t_raw` is not defined in our tests.

Comment (zou3519) [reply aten/src/ATen/native/Linear.cpp:50]: This makes sense to me. The problem is that if we go down the `add_` route we have `DualTensor[primal: Tensor, tangent: Tensor].add_(DualTensor[primal: Tensor, tangent: CCTensor]` and there is way to change the tangent on the LHS to be a CCTensor.

For this PR, I want to figure out if the way we are checking if a Tensor has forward-mode AD is legit (@soulitzer ?).

Then, either now or in the future, we should consider if all dual tensors should return True for isTensorSubclassLike  (I think they should -- a dual tensor is a "Tensor subclass" in spirit). Thoughts, @kshitij12345, @soulitzer, @albanD?

File: aten/src/ATen/native/Linear.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/Linear.cpp:51]: Should use https://github.com/pytorch/pytorch/blob/855c4eb0514cd8919b2f91a840c950304bbece66/torch/csrc/autograd/functions/utils.h#L90-L92 as the check

Comment (soulitzer) [reply aten/src/ATen/native/Linear.cpp:51]: nit: `bias->_fw_grad(/*level*/ 0).defined()), but yeah that should be fine!

File: aten/src/ATen/native/Linear.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/Linear.cpp:50]: > More precisely, the case we want to avoid is here is in-placing a subclassed tangent into a non-subclassed tangent right? 

Yes

> Could we just update the is_subclass check to also check whether the tangent is subclass then?

Seems reasonable to me

Comment (soulitzer) [reply aten/src/ATen/native/Linear.cpp:50]: Shouldn't need to be recursive because we disallow tangents from having tangents themselves at the same level (we also don't support multiple levels).

File: aten/src/ATen/native/Linear.cpp (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/Linear.cpp]: Done.

**Review by soulitzer — APPROVED** [review]
Thanks!
